File: internal/report/share_hash.go
package report

import (
	"net/url"
	"strings"
)

// ShareHash is the static report URL fragment contract mirrored by static_report.js.
// Format: #<focus>[?q=&spec=&scenario=&failSteps=1]
// Legacy: #fail-steps (implies overview + failSteps).
type ShareHash struct {
	Focus     string
	Query     string
	Spec      string
	Scenario  string
	FailSteps bool
}

func normalizeShareVerdict(v string) string {
	switch strings.ToLower(strings.TrimSpace(v)) {
	case "pass", "fail", "skip":
		return strings.ToLower(strings.TrimSpace(v))
	default:
		return "all"
	}
}

// ParseShareHash parses a location.hash (with or without leading #).
func ParseShareHash(raw string) ShareHash {
	out := ShareHash{Focus: "overview", Spec: "all", Scenario: "all"}
	input := strings.TrimPrefix(strings.TrimSpace(raw), "#")
	if input == "" {
		return out
	}
	head, qs, hasQS := strings.Cut(input, "?")
	failSteps := false
	focus := head
	if head == "fail-steps" || strings.HasPrefix(head, "fail-steps&") || strings.HasPrefix(head, "fail-steps/") {
		focus = "overview"
		failSteps = true
		if strings.HasPrefix(head, "fail-steps&") {
			extra := strings.TrimPrefix(head, "fail-steps&")
			if hasQS {
				qs = extra + "&" + qs
			} else {
				qs = extra
				hasQS = true
			}
		}
	}
	if focus == "" {
		focus = "overview"
	}
	out.Focus = focus
	out.FailSteps = failSteps
	if hasQS && qs != "" {
		params, err := url.ParseQuery(qs)
		if err == nil {
			if params.Has("q") {
				out.Query = params.Get("q")
			}
			if params.Has("query") {
				out.Query = params.Get("query")
			}
			if sp := params.Get("spec"); sp != "" {
				out.Spec = normalizeShareVerdict(sp)
			}
			if sc := params.Get("scenario"); sc != "" {
				out.Scenario = normalizeShareVerdict(sc)
			} else if sc := params.Get("scn"); sc != "" {
				out.Scenario = normalizeShareVerdict(sc)
			}
			fs := params.Get("failSteps")
			if fs == "" {
				fs = params.Get("fail-steps")
			}
			switch strings.ToLower(fs) {
			case "1", "true", "yes":
				out.FailSteps = true
			case "0", "false", "no":
				out.FailSteps = false
			}
		}
	}
	out.Spec = normalizeShareVerdict(out.Spec)
	out.Scenario = normalizeShareVerdict(out.Scenario)
	return out
}

// FormatShareHash builds a fragment without leading #.
func FormatShareHash(h ShareHash) string {
	focus := strings.TrimSpace(h.Focus)
	if focus == "" || focus == "overview" {
		focus = "overview"
	}
	params := url.Values{}
	q := strings.TrimSpace(h.Query)
	if q != "" {
		params.Set("q", q)
	}
	if sp := normalizeShareVerdict(h.Spec); sp != "all" {
		params.Set("spec", sp)
	}
	if sc := normalizeShareVerdict(h.Scenario); sc != "all" {
		params.Set("scenario", sc)
	}
	if h.FailSteps {
		params.Set("failSteps", "1")
	}
	enc := params.Encode()
	if enc == "" {
		return focus
	}
	return focus + "?" + enc
}

File: desktop/electron/share-hash.js
'use strict';

/**
 * Mirror of internal/report ParseShareHash / FormatShareHash for Desktop.
 * Fragment: #<focus>[?q=&spec=&scenario=&failSteps=1]
 */

function normalizeVerdict(v) {
  const s = String(v || '').trim().toLowerCase();
  return s === 'pass' || s === 'fail' || s === 'skip' ? s : 'all';
}

function parseShareHash(raw) {
  const out = { focus: 'overview', query: '', spec: 'all', scenario: 'all', failSteps: false };
  let input = String(raw || '').trim().replace(/^#/, '');
  if (!input) return out;
  let head = input;
  let qs = '';
  const qIdx = input.indexOf('?');
  if (qIdx >= 0) {
    head = input.slice(0, qIdx);
    qs = input.slice(qIdx + 1);
  }
  let failSteps = false;
  let focus = head;
  if (head === 'fail-steps' || head.startsWith('fail-steps&') || head.startsWith('fail-steps/')) {
    focus = 'overview';
    failSteps = true;
    if (head.startsWith('fail-steps&')) {
      const extra = head.slice('fail-steps&'.length);
      qs = qs ? `${extra}&${qs}` : extra;
    }
  }
  out.focus = focus || 'overview';
  out.failSteps = failSteps;
  if (qs) {
    const params = new URLSearchParams(qs);
    if (params.has('q')) out.query = String(params.get('q') || '');
    if (params.has('query')) out.query = String(params.get('query') || out.query);
    if (params.get('spec')) out.spec = normalizeVerdict(params.get('spec'));
    const sc = params.get('scenario') || params.get('scn');
    if (sc) out.scenario = normalizeVerdict(sc);
    const fs = params.get('failSteps') || params.get('fail-steps') || '';
    if (['1', 'true', 'yes'].includes(String(fs).toLowerCase())) out.failSteps = true;
    if (['0', 'false', 'no'].includes(String(fs).toLowerCase())) out.failSteps = false;
  }
  out.spec = normalizeVerdict(out.spec);
  out.scenario = normalizeVerdict(out.scenario);
  return out;
}

function formatShareHash(h = {}) {
  let focus = String(h.focus || 'overview').trim() || 'overview';
  if (focus === 'overview') focus = 'overview';
  const params = new URLSearchParams();
  const q = String(h.query || '').trim();
  if (q) params.set('q', q);
  const spec = normalizeVerdict(h.spec);
  const scenario = normalizeVerdict(h.scenario);
  if (spec !== 'all') params.set('spec', spec);
  if (scenario !== 'all') params.set('scenario', scenario);
  if (h.failSteps) params.set('failSteps', '1');
  const enc = params.toString();
  return enc ? `${focus}?${enc}` : focus;
}

/**
 * Build a report URL hash from Desktop outline filter state.
 * Fail verdict also enables failSteps so the static tree matches the outline.
 */
function reportOpenHashFromOutline(outline = {}) {
  const query = String(outline.query || '').trim();
  const verdict = normalizeVerdict(outline.verdict);
  return formatShareHash({
    focus: 'overview',
    query,
    scenario: verdict,
    failSteps: verdict === 'fail',
  });
}

function appendShareHash(url, hash) {
  const base = String(url || '');
  if (!base) return base;
  const frag = String(hash || '').replace(/^#/, '');
  if (!frag || frag === 'overview') {
    // Still attach bare overview when caller wants a stable hash target.
    if (frag === 'overview' && !base.includes('#')) return `${base}#overview`;
    return base;
  }
  const without = base.split('#')[0];
  return `${without}#${frag}`;
}

module.exports = {
  parseShareHash,
  formatShareHash,
  reportOpenHashFromOutline,
  appendShareHash,
  normalizeVerdict,
};
